### PR TITLE
RATIS-2158. Let the snapshot sender and receiver use a new digester each time

### PR DESCRIPTION
**Pull request by 133tosakarin**
see https://issues.apache.org/jira/browse/RATIS-2158

### CODE REVIEW

**Comment by OneSizeFitsQuorum** [comment]
@szetszwo PTAL

File: ratis-server/src/main/java/org/apache/ratis/server/storage/SnapshotManager.java [hunk 1]
```diff
@@ -87,7 +89,7 @@ private FileChannel open(FileChunkProto chunk, File tmpSnapshotFile) throws IOEx
       }
       // create the temp snapshot file and put padding inside
       out = FileUtils.newFileChannel(tmpSnapshotFile, StandardOpenOption.WRITE, StandardOpenOption.CREATE);
-      digester = MD5Hash.newDigester();
+      digester.set(MD5Hash.getDigester());
```
Comment (szetszwo) [inline ratis-server/src/main/java/org/apache/ratis/server/storage/SnapshotManager.java:92]: @133tosakarin , why changing back to `getDigester`?

File: ratis-server/src/main/java/org/apache/ratis/server/storage/SnapshotManager.java [hunk 2]
```diff
@@ -62,7 +63,7 @@ public class SnapshotManager {
   private final Supplier<File> snapshotDir;
   private final Supplier<File> snapshotTmpDir;
   private final Function<FileChunkProto, String> getRelativePath;
-  private MessageDigest digester;
+  private final AtomicReference<MessageDigest> digester;
```
Comment (szetszwo) [inline ratis-server/src/main/java/org/apache/ratis/server/storage/SnapshotManager.java:66]: How about simply add a `volatile`?

**Review by szetszwo — COMMENTED** [review]
@133tosakarin , @OneSizeFitsQuorum , on a second thought, we may not need any synchronization.    The code must be already synchronized.  If not, two different threads accessing `digester` can change order.   Then the final `digester.digest()` will be different.

The reason is similar to `OutputStream`, whose subclasses (e.g. `FileOutputStream`) usually are NOT thread-safe.  If there is a need, the caller should synchronize it; see https://stackoverflow.com/questions/8422278/write-to-fileoutputstream-from-multiple-threads-in-java

If we really want to make a change, how about simply add `volatile`?

File: ratis-server/src/main/java/org/apache/ratis/server/storage/SnapshotManager.java (hunk 1 above)
Comment (133tosakarin) [reply ratis-server/src/main/java/org/apache/ratis/server/storage/SnapshotManager.java]: Sorry, I mistakenly thought that every time I use DIGESTER_FACTORY (), newDigest () will be called, so I want to try changing it to getDigester ().

I have changed back to newDigester.

File: ratis-server/src/main/java/org/apache/ratis/server/storage/SnapshotManager.java (hunk 2 above)
Comment (133tosakarin) [reply ratis-server/src/main/java/org/apache/ratis/server/storage/SnapshotManager.java]: ok

File: ratis-server/src/main/java/org/apache/ratis/server/storage/SnapshotManager.java [hunk 3]
```diff
@@ -62,7 +63,7 @@ public class SnapshotManager {
   private final Supplier<File> snapshotDir;
   private final Supplier<File> snapshotTmpDir;
   private final Function<FileChunkProto, String> getRelativePath;
-  private MessageDigest digester;
+  private volatile AtomicReference<MessageDigest> digester;
```
Comment (szetszwo) [inline ratis-server/src/main/java/org/apache/ratis/server/storage/SnapshotManager.java:66]: @133tosakarin , sorry, I mean that don't use `AtomicReference`, i.e.
```java
 private volatile MessageDigest digester;
```

**Review by szetszwo — APPROVED** [review]
+1 the change looks good.

Let's also update the title?

**Comment by OneSizeFitsQuorum** [comment]
Hi, @szetszwo @133tosakarin ,
1. Maybe we need to create a new one for [sender](https://github.com/apache/ratis/blob/master/ratis-server/src/main/java/org/apache/ratis/server/storage/FileChunkReader.java#L59) as we already saw that the reset on the receiver side can sometimes fail, and there is no significant performance cost to creating a new digester?
2. Thanks for the reminder. Now that we have synchronized, we probably don't need to add another judgment; adding volatile would cause a sonar bug. Maybe we can undo this change.
3. It seems the ci failure is same as this problem discussed in [mail list](https://lists.apache.org/thread/m4r34p2y79z4rvy5vyc983ty06xxt92x), then we can merge this pr first and create a new issue for this problem for master.
image

**Comment by szetszwo** [comment]
>  t seems the ci failure is same as this problem discussed in [mail list](https://lists.apache.org/thread/m4r34p2y79z4rvy5vyc983ty06xxt92x), ...

@OneSizeFitsQuorum , if the failure is related to `ReferenceCount` or zero-copy, let ignore it for the moment.  We will fix it separately.

**Review by szetszwo — COMMENTED** [review]
@133tosakarin , thanks for the update!  Please see the comment inlined.

File: ratis-common/src/main/java/org/apache/ratis/io/MD5Hash.java [hunk 4]
```diff
@@ -34,7 +34,9 @@ public class MD5Hash {
 
   public static MessageDigest newDigester() {
     try {
-      return MessageDigest.getInstance("MD5");
+      MessageDigest digester = MessageDigest.getInstance("MD5");
+      digester.reset();
```
[the remark this answers is not in the document]
Comment (133tosakarin) [reply ratis-common/src/main/java/org/apache/ratis/io/MD5Hash.java]: all right

**Review by OneSizeFitsQuorum — APPROVED** [review]
LGTM

**Review by szetszwo — APPROVED** [review]
+1 the change looks good.

**Pull request merged**